File: deploy-migrations.js
#!/usr/bin/env node

/**
 * Deploy Supabase migrations using Supabase JS client
 * Usage: node deploy-migrations.js
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

// Configuration from environment
const SUPABASE_URL = process.env.VITE_SUPABASE_URL || 'https://cadavbabblukuabioekc.supabase.co';
const SUPABASE_KEY = process.env.VITE_SUPABASE_ANON_KEY;

if (!SUPABASE_URL || !SUPABASE_KEY) {
  console.error('❌ Missing SUPABASE_URL or SUPABASE_KEY in environment');
  console.error('   Please set VITE_SUPABASE_URL and VITE_SUPABASE_ANON_KEY');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
const MIGRATIONS_DIR = path.join(__dirname, 'supabase', 'migrations');

/**
 * Execute SQL via Supabase REST API
 */
async function executeSql(sql) {
  try {
    // Use the RPC method to execute SQL
    // Note: This requires a database function to be created
    // For now, we'll use a workaround with the raw API
    const response = await fetch(`${SUPABASE_URL}/rest/v1/rpc/execute_sql`, {
      method: 'POST',
      headers: {
        'apikey': SUPABASE_KEY,
        'Authorization': `Bearer ${SUPABASE_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ sql_text: sql }),
    });

    if (!response.ok) {
      const error = await response.text();
      console.error(`   ❌ Error: ${error}`);
      return false;
    }

    console.log(`   ✅ Success`);
    return true;
  } catch (error) {
    console.error(`   ❌ Exception: ${error.message}`);
    return false;
  }
}

/**
 * Main deployment logic
 */
async function deploy() {
  console.log('🚀 Deploying Supabase migrations...\n');

  // List migrations
  const migrations = fs.readdirSync(MIGRATIONS_DIR)
    .filter(f => f.endsWith('.sql'))
    .sort();

  if (migrations.length === 0) {
    console.error('❌ No migrations found');
    return false;
  }

  console.log(`📁 Found ${migrations.length} migrations:`);
  migrations.forEach(m => console.log(`   - ${m}`));
  console.log();

  // Execute each migration
  let allSuccess = true;
  for (const migration of migrations) {
    const filePath = path.join(MIGRATIONS_DIR, migration);
    console.log(`📄 Executing ${migration}...`);

    try {
      const sql = fs.readFileSync(filePath, 'utf8');
      const success = await executeSql(sql);
      if (!success) {
        allSuccess = false;
      }
    } catch (error) {
      console.error(`   ❌ Failed to read file: ${error.message}`);
      allSuccess = false;
    }
  }

  console.log();
  if (allSuccess) {
    console.log('✅ All migrations deployed successfully!');
    console.log('\n📋 Verify in Dashboard:');
    console.log('   1. Go to https://app.supabase.com');
    console.log('   2. Select project: cadavbabblukuabioekc');
    console.log('   3. Click "Table Editor"');
    console.log('   4. Check that episodes, axes, questions exist');
  } else {
    console.log('⚠️  Some migrations failed');
  }

  return allSuccess;
}

// Run deployment
deploy()
  .then(success => process.exit(success ? 0 : 1))
  .catch(error => {
    console.error('❌ Unexpected error:', error);
    process.exit(1);
  });

File: deploy_migrations.py
#!/usr/bin/env python3
"""
Deploy Supabase migrations using REST API
"""
import json
import requests
import sys
from pathlib import Path

# Configuration
SUPABASE_URL = "https://cadavbabblukuabioekc.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNhZGF2YmFiYmx1a3VhYmlvZWtjIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjU3OTc3MDksImV4cCI6MjA4MTM3MzcwOX0.HOVGnZDLUx-taLUe_rqYf7Z-cRUmDFj783SIPLvkECk"
MIGRATIONS_DIR = Path("/workspaces/constellation-chronicle/supabase/migrations")

headers = {
    "apikey": SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type": "application/json",
}

def execute_sql(sql: str) -> bool:
    """Execute SQL via Supabase REST API"""
    # Escape newlines and quotes in SQL
    sql_escaped = sql.replace('"', '\\"').replace('\n', ' ')
    
    payload = {
        "query": sql
    }
    
    try:
        response = requests.post(
            f"{SUPABASE_URL}/rest/v1/rpc/exec_sql",
            json=payload,
            headers=headers,
            timeout=30
        )
        print(f"Status: {response.status_code}")
        if response.status_code >= 400:
            print(f"Error: {response.text}")
            return False
        print(f"Response: {response.text[:200]}")
        return True
    except Exception as e:
        print(f"Exception: {e}")
        return False

def main():
    """Deploy all migrations"""
    migrations = sorted(MIGRATIONS_DIR.glob("*.sql"))
    
    if not migrations:
        print("❌ No migrations found")
        return False
    
    print(f"📁 Found {len(migrations)} migrations:")
    for m in migrations:
        print(f"   - {m.name}")
    
    print("\n🚀 Deploying migrations...")
    
    all_success = True
    for migration_file in migrations:
        print(f"\n📄 Executing {migration_file.name}...")
        
        sql = migration_file.read_text()
        
        # For large migrations, split by statements
        # This is a simple approach - may need refinement
        if execute_sql(sql):
            print(f"   ✅ Success")
        else:
            print(f"   ❌ Failed")
            all_success = False
    
    if all_success:
        print("\n✅ All migrations deployed successfully!")
        return True
    else:
        print("\n⚠️  Some migrations failed")
        return False

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)
